### DISCUSSION TITLE
Installing Python sources outside the site-packages

### DISCUSSION THREAD

**Discussion by igo95862** [mesonbuild/meson#13132] [Q&A]
I want to install the Python files outside the site-packages directories so that it is not Python version dependent.

My first idea was to modify the `python.purelibdir` and `python.platlibdir` options, however, it is impossible to programatically modify the options unless they are in subproject

I tried setting `subdir` to an absolute path but then it no longer compiles byte code.

Instead I found this trick:

```
py_installation.install_sources(
    source_files,
    subdir : fs.relative_to('/' / python_packages_dir, py_installation.get_path('purelib')),
)
```

You get both byte code compilation and it is installed outside the site-packages. However, it is kind of a hack.

Maybe there is a better solution?

**Comment by eli-schwartz** [comment]
Why do you want it to be independent of the python version?

Note that on e.g. Debian, the default installation location for install_sources is already "python3/dist-packages" which is independent of version

That's actually a bit of a problem since you can install the interpreter for python 3.8 and python 3.10 and then install a system package for a python 3.10 module, which is then importable in python 3.8 but is broken and fails to work (despite which, pip install thinks it's already installed and doesn't try to install a compatible copy).

You cannot actually use a module with multiple versions of python unless it's a pure library. For applications (such as bubbblejail??) you are limited to whichever version of python is in the script shebang -- and also whichever version of python has your python dependencies installed.

**Reply by igo95862** [reply]
> Why do you want it to be independent of the python version?

Because main distribution mode for bubblejail is AUR for which packages don't automatically update together with system. So when Python gets updated if you don't use version independent modules directory your entire desktop suddenly will become non-functional as all the browsers you sandboxed will stop working as bubblejail will fail to launch.

I currently install the packages to `/usr/lib/bubblejail/python_packages` and my launch script manipulates `sys.path` to include this directory. This setup can survive the Python upgrade and the only downside is that the bytecode will no longer be used. I created my own bytecode compile script before meson had ability to compile bytecode but I was looking in to switching to meson's native bytecode compilation. However, lack of ability to compile bytecode outside the `site-packages` is a major road block.

> You cannot actually use a module with multiple versions of python unless it's a pure library.

Even if it has C modules you can use the limited API which makes them compatible across multiple versions.

**Reply by eli-schwartz** [reply]
When I say a pure library I don't mean a pure .py project, I mean it's purely a library as opposed to an application.

This is distinct from whether .py files are actually compatible with multiple versions of python -- they frequently are not and then require source code changes to fix that.

**Reply by igo95862** [reply]
I've never encountered that but I keep the number of dependent libraries to a minimum and I only expect the Python to upgraded forwards. The standard library is pretty good at having forwards compatibility.

**Reply by eli-schwartz** [reply]
> Because main distribution mode for bubblejail is AUR for which packages don't automatically update together with system. So when Python gets updated if you don't use version independent modules directory your entire desktop suddenly will become non-functional as all the browsers you sandboxed will stop working as bubblejail will fail to launch.

What about non-Arch distros? :)

> I currently install the packages to `/usr/lib/bubblejail/python_packages` and my launch script manipulates `sys.path` to include this directory. This setup can survive the Python upgrade and the only downside is that the bytecode will no longer be used.

The other downside is that if you ever run bubblejail as root, it will create files which are untracked by pacman. Then the next time you build the AUR package against the new python version, it tries to install files that conflict with the untracked files, and pacman fatally aborts.

**Reply by eli-schwartz** [reply]
> I've never encountered that but I keep the number of dependent libraries to a minimum and I only expect the Python to upgraded forwards. The standard library is pretty good at having forwards compatibility.

The standard library is actually fairly prone to breaking backwards compatibility with existing code. The trick is that they don't do so intentionally without good reason, and when they do, they provide a deprecation notice for a couple minor versions.

I've still seen enough cases where internal refactoring e.g. caused a file descriptor to be closed automatically for you, and suddenly code that assumed it had to be manually closed would fail with a traceback when trying to do so (in a bugfix release of CPython, by the way!).

Minor versions of python come with a list of deprecations and removals and probably you just don't happen to use those corners of the language.

There are also some widely publicized expected breakages in the name of progress which libraries *do* fix by adding support for the newer python versions, but where Debian oldstale still provides ancient versions of that library which simply lack the fixes; this combines like oil and fire with the fact that Debian also provides backported versions of newer python interpreters that aren't tested to work with the libraries they package, but automatically get used with their unversioned sys.path.

**Reply by igo95862** [reply]
> What about non-Arch distros? :)

This setup also works for any other distro. (`python_packages_dir = get_option('libdir') / 'bubblejail/python_packages/'
`) For distro maintainers there is a build option to use the `py_installation.get_path('purelib')` which I know the Alpine Linux uses.

> The other downside is that if you ever run bubblejail as root

You aren't supposed to run it as root as it is completely unprivileged. It expects to be run in the desktop environment as without `XDG_RUNTIME_DIR` it will fail to run. Sometimes people run it as root probably because of how `firejail` works so I should probably add the abort to launch script then it runs as root. There is also [sys.dont_write_bytecode](https://docs.python.org/3/library/sys.html#sys.dont_write_bytecode) which can be manipulated to disable bytecode or the launch script shebang can include the `-B` to disable bytecode writing.

**Reply by eli-schwartz** [reply]
> This setup also works for any other distro.

It's an unusual complication, I'd say, for any distro that officially packages bubblejail. In particular it makes it harder to detect cases where the package should be rebuilt (you can't just check for packages which install files to a deprecated directory).

> You aren't supposed to run it as root as it is completely unprivileged.

Bytecode compilation occurs at import time, which also means when running the program as root in order to see the `--help` text.

**Reply by igo95862** [reply]
> Bytecode compilation occurs at import time, which also means when running the program as root in order to see the `--help` text.

I can manipulate the `sys.dont_write_bytecode` before importing any bubblejail's modules.

Anyway this is a minor detail. The important part is that I want to make bubblejail survive Python upgrades.

If meson does not have a good way of installing the Python sources outside the `purelib` I will continue using my bytecode compiler. I also plan on using a subproject with [my Python namespaces library](https://github.com/igo95862/python-lxns). It looks like I manipulate the subproject settings to modify the `platlibdir` of the subproject. Not sure how the byte code compilation will work.

An alternative solution I can think of is recording the build time `purelib` in the launch script and then checking if it is still in the `sys.path` before importing bubblejail modules. If it is no longer there that means the Python version was changed so insert it back and maybe emit a warning that user should rebuild their bubblejail.  Not sure how vendoring the subprojects will work as I don't want to install the vendored subprojects inside the `purelib` or they might conflict. Maybe record an extra `sys.path` in the launch script with the vendored modules path.

**Reply by igo95862** [reply]
I've upgraded to Python 3.12 and bubblejail still works without rebuilding meaning my approach of installing in to a version independent directory is validated.

**Reply by eli-schwartz** [reply]
- that doesn't solve the bytecode issue
- it definitely doesn't validate the approach for other software, which means I'm hesitant to provide a mechanism other software might be tempted to use also
  - it's not guaranteed to work for bubblejail either when upgrading to 3.13, just because it happened to work for 3.12

**Reply by igo95862** [reply]
So far it worked 3.9->3.10->3.11->3.12.

Anyway it does not look like I can control subprojects purelib so I will instead use `error()` if the `-Dpython.purelib` does not match the version independent directory I want to use. AUR users will be able to do a `git pull` to update PKGBUILD with a correct `-Dpython.purelib`.

**Reply by igo95862** [reply]
I upgraded to Python 3.13 and everything still works.
